src/SqlBinding/SqlAsyncCollector.cs

```diff
@@ -90,55 +92,284 @@ public async Task FlushAsync(CancellationToken cancellationToken = default)
+                    ON 
+                        {primaryKeyMatchingQuery}
+                    WHEN MATCHED THEN
+                        UPDATE SET {columnMatchingQuery}
+                    WHEN NOT MATCHED THEN 
+                        INSERT ({string.Join(",", columnNames)}) VALUES ({string.Join(",", columnNames)})";
+            }
+
+            /// <summary>
+            /// Retrieve (relatively) static information of SQL Table like primary keys, column names, etc. 
+            /// in order to generate the MERGE portion of the upsert query.
+            /// This only needs to be generated once and can be reused for subsequent upserts.
+            /// </summary>
+            /// <param name="sqlConnection">Connection with which to query SQL against</param>
+            /// <param name="fullName">Full name of table, including schema (if exists).</param>
+            /// <returns>TableInformation object containing primary keys, column types, etc.</returns>
+            public async static Task<TableInformation> RetrieveTableInformationAsync(SqlConnection sqlConnection, string fullName)
+            {
+                SqlBindingUtilities.GetTableAndSchema(fullName, out string schema, out string tableName);
+
+                // 1. Get all column names and types
+                var columnDefinitionsFromSQL = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+                try
+                {
+                    await sqlConnection.OpenAsync();
+                    SqlCommand cmdColDef = new SqlCommand(GetColumnDefinitionsQuery(schema, tableName), sqlConnection);
+                    using (SqlDataReader rdr = await cmdColDef.ExecuteReaderAsync())
                     {
-                        builder = new SqlCommandBuilder(dataAdapter);
-                        _commandBuilders.TryAdd(key, builder);
+                        while (rdr.Read())
+                        {
+                            columnDefinitionsFromSQL.Add(rdr[ColumnName].ToString(), rdr[ColumnDefinition].ToString());
+                        }
                     }
-                    else
+                }
+                catch (Exception ex)
+                {
+                    // Throw a custom error so that it's easier to decipher.
+                    string message = $"Encountered exception while retrieving columns definition for table '{tableName}.' Cannot generate upsert command without them.";
+                    throw new InvalidOperationException(message, ex);
+                }
+                finally
+                {
+                    await sqlConnection.CloseAsync();
+                }
+
+                // 2. Ensure the POCO fields match the SQL column names. If not, throw. (TODO: is this expected behavior?)
+                IEnumerable<string> columnNamesFromPOCO = typeof(T).GetMembers().Where(f => f.MemberType == MemberTypes.Property || f.MemberType == MemberTypes.Field).Select(f => f.Name);
```

Comment:
I agree, this implementation feels somewhat ridiculous to do manually, rather than some existing library. However, in my limited experience with EF, I haven't been very impressed by its performance. From what I understand, its version of merge (`AddOrUpdate`) requires two round-trips to the database, rather than just one. I'm not sure if there are better alternatives to EF (@yorek - do you know of any?)

Reply:
I would suggest to use Dapper and something like Automapper (https://automapper.org/).